src/test/coinsviewcacheasync_tests.cpp

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <coins.h>
+#include <coinsviewcacheasync.h>
+#include <crypto/common.h>
+#include <primitives/block.h>
+#include <primitives/transaction.h>
+#include <primitives/transaction_identifier.h>
+#include <test/util/random.h>
+#include <test/util/setup_common.h>
+#include <uint256.h>
+#include <util/hasher.h>
+
+#include <boost/test/unit_test.hpp>
+
+#include <cstdint>
+#include <ranges>
+#include <unordered_set>
+
+BOOST_AUTO_TEST_SUITE(coinsviewcacheasync_tests)
+
+struct NoAccessCoinsView : CCoinsView {
+    std::optional<Coin> GetCoin(const COutPoint&) const override { abort(); }
+};
+
+struct CoinsViewCacheAsyncTest : BasicTestingSetup {
+private:
+    std::unique_ptr<CoinsViewCacheAsync> m_async_cache{nullptr};
+    std::unique_ptr<CBlock> m_block{nullptr};
+
+    CBlock CreateBlock(int32_t num_txs)
+    {
+        CBlock block;
+        CMutableTransaction coinbase;
+        coinbase.vin.emplace_back();
+        block.vtx.push_back(MakeTransactionRef(coinbase));
+
+        Txid prevhash{Txid::FromUint256(uint256(1))};
+
+        for (const auto i : std::views::iota(1, num_txs)) {
+            CMutableTransaction tx;
+            Txid txid;
+            if (i % 3 == 0) {
+                txid = Txid::FromUint256(uint256(i));
+            } else if (i % 3 == 1) {
+                txid = prevhash;
+            } else {
+                // Test shortid collisions
+                const uint64_t shorttxid{prevhash.ToUint256().GetUint64(0)};
+                uint256 u(i);
+                WriteLE64(u.data(), shorttxid);
+                txid = Txid::FromUint256(u);
+            }
+            tx.vin.emplace_back(COutPoint(txid, 0));
```

Comment:
nit: emplace doesn't necessarily need the class name

```suggestion
            tx.vin.emplace_back(txid, 0);
```